Initial Contour Controller Implementation

danehans opened the pull request:
- Moves `main()` to cmd/operator pkg. `cmd/operator/start.go `: Exposes a flag to set the Contour image used by the operator.
- Singularizes the controller pkg.
- `controller/contour/controller.go`: Initial controller implementation that watches the contour CRD and a) finalizes the resource and b) creates the contour namespace if it does not exist.
- `util/retryableerror/retryableerror.go`: Utility function for retrying reconciliation errors.
- `util/slice/slice.go`: Utility function to check if a given slice of strings contains the provided string.

Signed-off-by: Daneyon Hansen <daneyonhansen@gmail.com>

danehans commented:
/assign @jpeach 
/cc @youngnick @Miciah

jpeach commented on cmd/operator/start.go:
```diff
@@ -76,9 +83,9 @@ func main() {
+	setupLog.Info("starting contour-operator")
 	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "problem running manager")
+		setupLog.Error(err, "problem running contour-operator")
```
nit: "failed to start contour-operator" would be a marginally better error

danehans replied:
ack

jpeach commented on controller/contour/controller.go:
```diff
@@ -0,0 +1,123 @@
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=operator.projectcontour.io,resources=contours,verbs=get;list;watch;create;update;patch;delete
```
Do we need `create;update;patch;delete` access on the main resource?

danehans replied:
I'll change to `get;list;watch;update`. We'll most likely have future use cases that will require the additional verbs, but the they can be updated as needed.

jpeach commented on controller/contour/namespace.go:
```diff
@@ -0,0 +1,49 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
```
Group imports by stdlib, then everything else.

jpeach commented on util/retryableerror/retryableerror_test.go:
```diff
@@ -0,0 +1,79 @@
+		} else if gotRetryable && retryable.After() != test.expectAfter {
+			t.Errorf("%q: expected after %v, got %v: %v", test.name, test.expectAfter, retryable.After(), err)
+		}
+		if _, gotAggregate := err.(utilerrors.Aggregate); gotAggregate != test.expectAggregate {
```
Nit: `testify/assert` is pretty good

danehans replied:
I created issue https://github.com/projectcontour/contour-operator/issues/13 to give this decision the proper treatment.

jpeach commented on util/retryableerror/retryableerror_test.go:
```diff
@@ -0,0 +1,79 @@
+			expectAfter:     time.Second * 15,
+		},
+	}
+	for _, test := range tests {
```
You can use subtests here:
```Go
cases := map[string]struct{ ... } { ... }
for name, case := range cases {
   t.Run(name, func(t *testing.T) {
      err := NewMaybeRetryableAggregate(case.errors)
      ...
   }
}
```

jpeach commented on controller/contour/controller.go:
```diff
@@ -0,0 +1,123 @@
+	}
+
+	// The contour is safe to process, so ensure current state matches desired state.
+	desired := true
```
I guess it's ok for now, but 

1. passing a bool is harder to read and understand than an enum
2. teardown probably needs to be done in the reverse order

danehans replied:
Yeah, I refactored based on https://github.com/projectcontour/contour-operator/pull/10#discussion_r481492143.

jpeach commented on controller/contour/namespace.go:
```diff
@@ -0,0 +1,49 @@
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: name}}
+	if err := r.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get namespace %s: %v", ns.Name, err)
```
This isn't an error if we are tearing down the namespace, though we wold want to update status in that case (maybe just add a TODO here?)

danehans replied:
ack, fixed.

danehans commented on controller/contour/controller.go:
```diff
@@ -0,0 +1,123 @@
+
+	// The contour is safe to process, so ensure current state matches desired state.
+	desired := true
+	if err := r.ensureContour(desired, contour); err != nil {
```
`ensureContour` is the func that will encapsulate all the lower-level funcs required to ensure dependent resources exist. A mirrored workflow will be created for removing contour and dependent resources. I've been working through the changes and will submit a new commit shortly. The pattern I'm trying to establish is similar to other operators and upstream k8s controllers. I'm open to feedback, so let me know your thoughts after I reviewing my upcoming commit.

danehans commented on controller/contour/controller.go:
```diff
@@ -0,0 +1,123 @@
+	"github.com/projectcontour/contour-operator/util/slice"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
```
I'm for stdlib, local, vendor. I'll clean up the imports with my upcoming commit.

jpeach replied:
Cool, I believe that `goimports` has a `-local` flag that may help with this; maybe we can add that to the lint config :)

danehans commented on Dockerfile:
```diff
@@ -10,18 +10,18 @@ COPY go.sum go.sum
 
 # Build
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o start start.go
```
ack

danehans commented:
@jpeach I believe commit `e83f7b1` addresses all your feedback except https://github.com/projectcontour/contour-operator/pull/10#discussion_r480759628. I've never implemented subtests like you mention. Do you have a Contour example to share or would you mind implementing after this merges so I can learn?

jpeach reviewed (approved):
LGTM
